### PR TITLE
Have `InvalidParam` support records

### LINKED ISSUE #2321: InvalidParam warning on Javadoc for Java record components
### Description of the problem / feature request:

Javadoc on Java record components are `@param` tags on the record itself. It's also possible to override the accessor methods, but that shouldn't be necessary. error-prone produces an `InvalidParam` warning here though, as it doesn't seem to understand the Java record syntax.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```java
/**
 * Example record.
 * @param name Name of the example record.
 */
public record ExampleRecord(String name) {
}
```

### What version of Error Prone are you using?

`2.6.0`

### PR DESCRIPTION
**Pull request by rickie**
Resolves #2321.

### CODE REVIEW

File: core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java [hunk 1]
```diff
@@ -101,6 +110,15 @@ public Description matchMethod(MethodTree methodTree, VisitorState state) {
     return Description.NO_MATCH;
   }
 
+  private static MethodTree getGeneratedConstructor(ClassTree classTree) {
+    return classTree.getMembers().stream()
```
Comment (rickie) [inline core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java:115]: For records in some cases the generated constructor could not be found by using `ASTHelpers#isGeneratedConstructor`. That's why it's implemented with the `contentEquals("<init>")` check. 

Perhaps that is a bigger issue we should solve. However, I think that is out of scope for this PR.

File: core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java [hunk 2]
```diff
@@ -101,6 +110,15 @@ public Description matchMethod(MethodTree methodTree, VisitorState state) {
...
+        .filter(MethodTree.class::isInstance)
+        .map(MethodTree.class::cast)
+        .findFirst()
+        .filter(tree -> tree.getName().contentEquals("<init>"))
+        .orElseThrow(() -> new IllegalStateException("Records must have a generated ctor"));
```
Comment (rickie) [inline core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java]: This error message could be better.

**Review by rickie — COMMENTED** [review]
Some context.

File: check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java [hunk 3]
```diff
@@ -798,6 +798,13 @@ private static boolean isFinal(Symbol symbol) {
     return (symbol.flags() & Flags.FINAL) == Flags.FINAL;
   }
 
+  private static final long RECORD_FLAG = 1L << 61;
```
Comment (rickie) [inline check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java:815]: We can't use `Flags.RECORD` yet, because that would fail with JDK 11...

Comment (dhoepelman) [inline check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java:815]:
```suggestion
  /**
   * Flag for a record class, its canonical constructor and components. Can be replaced by
   * com.sun.tools.javac.code.Flags.RECORD once support minimum JDK version is 14.
   */
  private static final long RECORD_FLAG = 1L << 61;
```

**Review by dhoepelman — APPROVED** [review]
Any timeline on merging this? 😇 We're peppering our code with `@SupressWarning("InvalidParam")` at the moment.

File: check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java (hunk 3 above)
Comment (rickie) [reply check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java:815]: Although I really like the suggestion, I used this one-liner because it's used in the Error Prone project twice like that and wanted to be consistent.

So I'd lean towards leaving it as-is. (When we use this in `error-prone-support` we should definitely use this as Javadoc.

Comment (rickie) [reply check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java:815]: Oh I misread, this is in the `ASTHelpers` of course... In that case, it is a good idea to add this. Will commit that.

File: check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java [hunk 4]
```diff
@@ -807,6 +807,17 @@ private static boolean isFinal(Symbol symbol) {
     return (symbol.flags() & Flags.FINAL) == Flags.FINAL;
   }
 
+  /**
+   * Flag for a record class, its canonical constructor and components. Can be replaced by {@code
```
Comment (Stephan202) [inline check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java]: W.r.t. `components`: not sure what is meant by that, but based on some tests with the debugger this does not include members. (The original documentation talks about the "state vector"; see below.) Maybe we should leave this out?

File: check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java [hunk 5]
```diff
@@ -807,6 +807,17 @@ private static boolean isFinal(Symbol symbol) {
...
+   * com.sun.tools.javac.code.Flags.RECORD} once the minimum JDK version is 14.
+   */
+  private static final long RECORD_FLAG = 1L << 61;
+
+  /** Returns true if the given {@link Symbol} is a record. */
+  public static boolean isRecord(Symbol symbol) {
+    return (symbol.flags() & RECORD_FLAG) == RECORD_FLAG;
+  }
```
Comment (Stephan202) [inline check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java]: The JDK 17 Javadoc for `com.sun.tools.javac.code.Flags.RECORD` says:
```
    /**
     * Flag to indicate that a class is a record. The flag is also used to mark fields that are
     * part of the state vector of a record and to mark the canonical constructor
     */
```

Not sure what the state vector is, but tests indicate that this method also returns `true` for the canonical constructor. We should either fix that or update the documentation.

File: core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java [hunk 6]
```diff
@@ -74,11 +75,19 @@ public final class InvalidParam extends BugChecker implements ClassTreeMatcher,
   public Description matchClass(ClassTree classTree, VisitorState state) {
     DocTreePath path = getDocTreePath(state);
     if (path != null) {
-      ImmutableSet<String> parameters = ImmutableSet.of();
+      // XXX: `isRecord`... should it also be able to handle `Tree`s?
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java]: :eyes:

Comment (rickie) [reply core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java]: I think we should defer it for now. We might want to improve this in a follow-up PR where we look at record handling in general.

File: core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java [hunk 7]
```diff
@@ -101,6 +110,15 @@ public Description matchMethod(MethodTree methodTree, VisitorState state) {
     return Description.NO_MATCH;
   }
 
+  private static MethodTree getGeneratedConstructor(ClassTree classTree) {
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java]: Given the current implementation:
```suggestion
  private static MethodTree getFirstConstructor(ClassTree classTree) {
```
This is not necessarily a generated constructor; see e.g. the `normalConstructor_record` test case.

That said, what we what, IIUC, is:
```suggestion
  private static MethodTree getCanonicalRecordConstructor(ClassTree classTree) {
```

But this does raise the question: can we reliable assume that the first constructor is always the canonical constructor? It might be better to explicitly test against `#isRecord`; will push a proposal.

File: check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java (hunk 4 above)
Comment (Stephan202) [reply check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java]: Upon further debugging, I found that a subset of record members _are_ flagged. I now better understand what is meant, so will update the documentation.

File: check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java (hunk 5 above)
Comment (Stephan202) [reply check_api/src/main/java/com/google/errorprone/util/ASTHelpers.java]: ^ Okay, IIUC the state vector comprises the fields that are populated by the canonical constructor.

File: core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java [hunk 8]
```diff
@@ -74,11 +75,19 @@ public final class InvalidParam extends BugChecker implements ClassTreeMatcher,
   public Description matchClass(ClassTree classTree, VisitorState state) {
     DocTreePath path = getDocTreePath(state);
     if (path != null) {
-      ImmutableSet<String> parameters = ImmutableSet.of();
+      // XXX: `isRecord`... should it also be able to handle `Tree`s?
+      ImmutableSet<String> parameters =
+          ASTHelpers.isRecord(ASTHelpers.getSymbol(classTree))
+              ? getGeneratedConstructor(classTree).getParameters().stream()
+                  .map(v -> v.getName().toString())
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/javadoc/InvalidParam.java]:
```suggestion
                  .map(p -> p.getName().toString())
```

**Review by Stephan202 — APPROVED** [review]
Rebased and added a commit.

Suggested commit message:
```
Have `InvalidParam` support records 

Fixes #2321.
```

**Comment by rickie** [comment]
Filed https://github.com/google/error-prone/pull/3437.

**Comment by rickie** [comment]
Fixed in https://github.com/google/error-prone/issues/2321.
